File: hodor/jira.py
"""Jira integration for Hodor code review.

Extracts Jira issue URLs from MR descriptions, fetches issue details,
and provides context for AI-powered code reviews.
"""

from __future__ import annotations

import base64
import logging
import os
import re
from typing import Any

import requests

logger = logging.getLogger(__name__)

# Regex to match Jira URLs like https://company.atlassian.net/browse/PROJECT-123
JIRA_URL_PATTERN = re.compile(
    r'https?://([a-zA-Z0-9-]+\.atlassian\.net)/browse/([A-Z][A-Z0-9]+-\d+)',
    re.IGNORECASE
)


class JiraAPIError(RuntimeError):
    """Raised when Jira API operations fail."""


def _get_jira_auth_header() -> dict[str, str] | None:
    """Build authentication header for Jira API.
    
    Uses JIRA_EMAIL and JIRA_API_KEY environment variables.
    Returns None if credentials are not configured.
    """
    email = os.getenv("JIRA_EMAIL")
    api_key = os.getenv("JIRA_API_KEY")
    
    if not email or not api_key:
        return None
    
    # Jira uses Basic auth with email:api_token
    credentials = f"{email}:{api_key}"
    encoded = base64.b64encode(credentials.encode()).decode()
    
    return {
        "Authorization": f"Basic {encoded}",
        "Content-Type": "application/json",
    }


def extract_jira_urls(text: str) -> list[tuple[str, str]]:
    """Extract Jira issue URLs from text.
    
    Args:
        text: Text to search (MR title, description, etc.)
        
    Returns:
        List of tuples: (jira_host, issue_key) e.g., ("company.atlassian.net", "EDR-1966")
    """
    if not text:
        return []
    
    matches = JIRA_URL_PATTERN.findall(text)
    # Deduplicate while preserving order
    seen = set()
    result = []
    for host, key in matches:
        key_upper = key.upper()
        if (host, key_upper) not in seen:
            seen.add((host, key_upper))
            result.append((host, key_upper))
    
    return result


def fetch_jira_issue(host: str, issue_key: str) -> dict[str, Any] | None:
    """Fetch Jira issue details via REST API.
    
    Args:
        host: Jira host (e.g., "company.atlassian.net")
        issue_key: Issue key (e.g., "EDR-1966")
        
    Returns:
        Issue data dict or None if fetch fails
    """
    auth_header = _get_jira_auth_header()
    if not auth_header:
        logger.warning("Jira credentials not configured (JIRA_EMAIL, JIRA_API_KEY)")
        return None
    
    url = f"https://{host}/rest/api/3/issue/{issue_key}"
    
    try:
        response = requests.get(
            url,
            headers=auth_header,
            timeout=10,
        )
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        logger.warning(f"Failed to fetch Jira issue {issue_key}: {e}")
        return None


def get_parent_issue(host: str, issue: dict[str, Any]) -> dict[str, Any] | None:
    """Fetch parent issue if the given issue is a subtask.
    
    Args:
        host: Jira host
        issue: Issue data from fetch_jira_issue
        
    Returns:
        Parent issue data or None
    """
    fields = issue.get("fields", {})
    
    # Check if this is a subtask
    issue_type = fields.get("issuetype", {})
    if not issue_type.get("subtask", False):
        return None
    
    # Get parent reference
    parent = fields.get("parent")
    if not parent:
        return None
    
    parent_key = parent.get("key")
    if not parent_key:
        return None
    
    logger.info(f"Fetching parent issue: {parent_key}")
    return fetch_jira_issue(host, parent_key)


def summarize_jira_issue(issue: dict[str, Any], is_parent: bool = False) -> str:
    """Format a single Jira issue for prompt context.
    
    Args:
        issue: Issue data from Jira API
        is_parent: Whether this is a parent issue
        
    Returns:
        Formatted markdown string
    """
    fields = issue.get("fields", {})
    key = issue.get("key", "Unknown")
    
    summary = fields.get("summary", "No summary")
    issue_type = fields.get("issuetype", {}).get("name", "Issue")
    status = fields.get("status", {}).get("name", "Unknown")
    priority = fields.get("priority", {}).get("name", "")
    
    # Get description (can be complex Atlassian Document Format)
    description = ""
    desc_field = fields.get("description")
    if desc_field:
        if isinstance(desc_field, str):
            description = desc_field
        elif isinstance(desc_field, dict):
            # Parse ADF (Atlassian Document Format) - simplified
            description = _extract_text_from_adf(desc_field)
    
    # Truncate description to a sane limit to prevent context explosion
    if len(description) > 5000:
        description = description[:4997] + "..."
    
    prefix = "**Parent Issue**" if is_parent else "**Linked Issue**"
    
    lines = [
        f"{prefix}: [{key}] {summary}",
        f"- Type: {issue_type}",
        f"- Status: {status}",
    ]
    
    if priority:
        lines.append(f"- Priority: {priority}")
    
    if description:
        lines.append(f"- Description: {description}")
    
    return "\n".join(lines)


def _extract_text_from_adf(adf: dict[str, Any]) -> str:
    """Extract plain text from Atlassian Document Format.
    
    Args:
        adf: ADF document structure
        
    Returns:
        Plain text content
    """
    text_parts = []
    
    def recurse(node: Any) -> None:
        if isinstance(node, dict):
            if node.get("type") == "text":
                text_parts.append(node.get("text", ""))
            for child in node.get("content", []):
                recurse(child)
        elif isinstance(node, list):
            for item in node:
                recurse(item)
    
    recurse(adf)
    return " ".join(text_parts).strip()


def build_jira_context(mr_title: str, mr_description: str) -> str:
    """Build Jira context section for the review prompt.
    
    Args:
        mr_title: MR title
        mr_description: MR description
        
    Returns:
        Formatted Jira context section (empty string if no Jira issues found)
    """
    # Combine title and description for URL extraction
    combined_text = f"{mr_title or ''}\n{mr_description or ''}"
    
    jira_urls = extract_jira_urls(combined_text)
    if not jira_urls:
        return ""
    
    logger.info(f"Found {len(jira_urls)} Jira issue(s) in MR")
    
    sections = []
    
    for host, issue_key in jira_urls:
        issue = fetch_jira_issue(host, issue_key)
        if not issue:
            continue
        
        # Add main issue
        sections.append(summarize_jira_issue(issue))
        
        # Check for parent
        parent = get_parent_issue(host, issue)
        if parent:
            sections.append(summarize_jira_issue(parent, is_parent=True))
    
    if not sections:
        return ""
    
    return "## Jira Context\n\n" + "\n\n".join(sections) + "\n"

File: src/jira.ts
import { logger } from "./utils/logger.js";

// Matches Jira URLs like https://company.atlassian.net/browse/PROJECT-123
const JIRA_URL_PATTERN = /https?:\/\/([a-zA-Z0-9-]+\.atlassian\.net)\/browse\/([A-Za-z][A-Za-z0-9]+-\d+)/gi;

export interface JiraUrlMatch {
  host: string;
  key: string;
}

interface AdfNode {
  type?: string;
  text?: string;
  content?: AdfNode[];
}

export interface JiraIssue {
  key?: string;
  fields?: {
    summary?: string;
    issuetype?: { name?: string; subtask?: boolean };
    status?: { name?: string };
    priority?: { name?: string };
    description?: string | AdfNode;
    parent?: { key?: string };
  };
}

function getJiraAuthHeader(): Record<string, string> | null {
  const email = process.env.JIRA_EMAIL;
  const apiKey = process.env.JIRA_API_KEY;
  if (!email || !apiKey) return null;

  const encoded = Buffer.from(`${email}:${apiKey}`, "utf-8").toString("base64");
  return {
    Authorization: `Basic ${encoded}`,
    "Content-Type": "application/json",
  };
}

/** Extract Jira issue URLs from text, deduplicated and normalized to uppercase keys. */
export function extractJiraUrls(text: string | null | undefined): JiraUrlMatch[] {
  if (!text) return [];

  const seen = new Set<string>();
  const result: JiraUrlMatch[] = [];
  for (const match of text.matchAll(JIRA_URL_PATTERN)) {
    const host = match[1];
    const key = match[2].toUpperCase();
    const dedupeKey = `${host}:${key}`;
    if (!seen.has(dedupeKey)) {
      seen.add(dedupeKey);
      result.push({ host, key });
    }
  }
  return result;
}

/** Fetch Jira issue details via REST API. Returns null if credentials are missing or the fetch fails. */
export async function fetchJiraIssue(host: string, issueKey: string): Promise<JiraIssue | null> {
  const authHeader = getJiraAuthHeader();
  if (!authHeader) {
    logger.warn("Jira credentials not configured (JIRA_EMAIL, JIRA_API_KEY)");
    return null;
  }

  const url = `https://${host}/rest/api/3/issue/${issueKey}`;
  try {
    const response = await fetch(url, {
      headers: authHeader,
      signal: AbortSignal.timeout(10_000),
    });
    if (!response.ok) {
      logger.warn(`Failed to fetch Jira issue ${issueKey}: HTTP ${response.status}`);
      return null;
    }
    return (await response.json()) as JiraIssue;
  } catch (err) {
    logger.warn(`Failed to fetch Jira issue ${issueKey}: ${err instanceof Error ? err.message : err}`);
    return null;
  }
}

/** Fetch the parent issue if the given issue is a subtask. */
export async function getParentIssue(host: string, issue: JiraIssue): Promise<JiraIssue | null> {
  const fields = issue.fields ?? {};
  if (!fields.issuetype?.subtask) return null;

  const parentKey = fields.parent?.key;
  if (!parentKey) return null;

  logger.info(`Fetching parent issue: ${parentKey}`);
  return fetchJiraIssue(host, parentKey);
}

/** Extract plain text from Atlassian Document Format (simplified). */
export function extractTextFromAdf(adf: unknown): string {
  const parts: string[] = [];

  function recurse(node: unknown): void {
    if (Array.isArray(node)) {
      for (const item of node) recurse(item);
      return;
    }
    if (node && typeof node === "object") {
      const obj = node as AdfNode;
      if (obj.type === "text" && typeof obj.text === "string") {
        parts.push(obj.text);
      }
      if (Array.isArray(obj.content)) {
        for (const child of obj.content) recurse(child);
      }
    }
  }

  recurse(adf);
  return parts.join(" ").trim();
}

/** Format a single Jira issue for prompt context. */
export function summarizeJiraIssue(issue: JiraIssue, isParent = false): string {
  const fields = issue.fields ?? {};
  const key = issue.key ?? "Unknown";
  const summary = fields.summary ?? "No summary";
  const issueType = fields.issuetype?.name ?? "Issue";
  const status = fields.status?.name ?? "Unknown";
  const priority = fields.priority?.name ?? "";

  let description = "";
  const descField = fields.description;
  if (typeof descField === "string") {
    description = descField;
  } else if (descField && typeof descField === "object") {
    description = extractTextFromAdf(descField);
  }

  // Truncate description to a sane limit to prevent context explosion
  if (description.length > 5000) {
    description = description.slice(0, 4997) + "...";
  }

  const prefix = isParent ? "**Parent Issue**" : "**Linked Issue**";
  const lines = [`${prefix}: [${key}] ${summary}`, `- Type: ${issueType}`, `- Status: ${status}`];

  if (priority) lines.push(`- Priority: ${priority}`);
  if (description) lines.push(`- Description: ${description}`);

  return lines.join("\n");
}

/** Build the Jira context section for the review prompt from an MR/PR title and description. */
export async function buildJiraContext(
  mrTitle?: string | null,
  mrDescription?: string | null,
): Promise<string> {
  const combinedText = `${mrTitle ?? ""}\n${mrDescription ?? ""}`;
  const jiraUrls = extractJiraUrls(combinedText);
  if (jiraUrls.length === 0) return "";

  logger.info(`Found ${jiraUrls.length} Jira issue(s) in MR`);

  const sections: string[] = [];
  for (const { host, key } of jiraUrls) {
    const issue = await fetchJiraIssue(host, key);
    if (!issue) continue;

    sections.push(summarizeJiraIssue(issue));

    const parent = await getParentIssue(host, issue);
    if (parent) sections.push(summarizeJiraIssue(parent, true));
  }

  if (sections.length === 0) return "";
  return "## Jira Context\n\n" + sections.join("\n\n") + "\n";
}
